pkg/controller/daemon/daemon_controller.go

```diff
@@ -535,6 +535,15 @@ func (dsc *DaemonSetsController) updatePod(old, cur interface{}) {
 		return
 	}
 
+	if curPod.DeletionTimestamp != nil {
+		// when a pod is deleted gracefully it's deletion timestamp is first modified to reflect a grace period,
+		// and after such time has passed, the kubelet actually deletes it from the store. We receive an update
+		// for modification of the deletion timestamp and expect an ds to create more replicas asap, not wait
+		// until the kubelet actually deletes the pod.
+		dsc.deletePod(curPod)
+		return
+	}
+
```

Comment:
i am confused how this fixes the issue ? Did you try with this fix and it solves the issue ?

Reply:
The reason why ds rolling update stucked is expectations not be satisfied. In this fix, we lower deletion expectation once pod's DeletionTimestamp is not nil to satisfy expectation and let syncloop resume.

Reply:
Before this fix, if a pod stucked in terminating, deamonset will never satisfy expectation.

Reply:
shouldnt lowering of expectation be done using dsc.expectations.DeletionObserved(dsKey) ?

Reply:
@janetkuo  could you help understand this fix ? Do we need a test ?

Reply:
> shouldnt lowering of expectation be done using dsc.expectations.DeletionObserved(dsKey) ?

of course you could just use dsc.expectations.DeletionObserved(dsKey)
but in this fix we could reuse the code in "deletePod" and keep code consistent with replicas-controller and job-controller. see https://github.com/kubernetes/kubernetes/pull/77773#issuecomment-492984652